### PR TITLE
API for External Processing filter

### PR DESCRIPTION
**Pull request by jenshu**
# Description

- Add the API for configuring the External Processing filter (an Enterprise-only feature). This can be configured globally in settings, or on an http listener, virtual host, or route.
- Moved around some merge util methods that can be shared.
- Translation will return an error if extproc is configured on a OSS gloo install.

# Checklist:

- [x] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/main/changelogutils) which references the issue that is resolved.
- [x] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: projects/gloo/api/v1/options.proto [hunk 1]
```diff
@@ -137,6 +138,17 @@ message HttpListenerOptions {
     // Enterprise-only: Settings for the cache server itself
     caching.options.gloo.solo.io.Settings caching = 17;
 
+    oneof ext_proc_config {
+        // Enterprise-only: Set to true to disable the External Processing filter for this listener.
+        // This can be overridden by child VirtualHostOptions or RouteOptions.
+        google.protobuf.BoolValue disable_ext_proc = 29;
```
Comment (ashishb-solo) [inline projects/gloo/api/v1/options.proto:144]: question: why this field is present in HttpListenerOptions but not VirtualHostOptions or RouteOptions?

File: projects/gloo/api/v1/enterprise/options/extproc/extproc.proto [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+import "google/protobuf/duration.proto";
+import "google/protobuf/struct.proto";
+import "google/protobuf/wrappers.proto";
+
+// Enterprise-only: Configuration for Envoy's [External Processing Filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/ext_proc_filter).
+// The External Processing filter allows for calling out to an external gRPC service at a specified
+// point within a HTTP filter chain. The external service may access and modify various parts of the
+// request or response, and may terminate processing.
+// Envoy's External Processing Filter is considered a work in progress and has an unknown security posture.
+// Only use this if you know what you are doing!
```
Comment (ashishb-solo) [inline projects/gloo/api/v1/enterprise/options/extproc/extproc.proto:28]: I feel like this language is a bit too informal

```suggestion
// Users should take care to understand the risks of using this extension before proceeding
```

File: projects/gloo/api/v1/options.proto (hunk 1 above)
Comment (jenshu) [reply projects/gloo/api/v1/options.proto:144]: the VirtualHostOptions/RouteOptions use a different message (extproc.options.gloo.solo.io.RouteSettings) which already has a disabled flag which envoy handles (lets you disable extproc on a vhost/route level, if it's enabled at the listener level).

the disable flag here is added so the user can potentially set default extproc settings globally (in Settings CR) but disable it for certain listeners

File: projects/gloo/api/v1/enterprise/options/extproc/extproc.proto [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+  // downstream client will be reset.
+  // With this parameter set to true, however, then if the gRPC stream is prematurely closed
+  // or could not be opened, processing continues without error.
+  google.protobuf.BoolValue failure_mode_allow = 3;
+
+  // Specifies default options for how HTTP headers, trailers, and bodies are
+  // sent.
+  .solo.io.envoy.extensions.filters.http.ext_proc.v3.ProcessingMode processing_mode = 4;
+
+  // Specifies the timeout for each individual message sent on the stream and
```
Comment (ashishb-solo) [inline projects/gloo/api/v1/enterprise/options/extproc/extproc.proto:49]: i'm having a little trouble understanding this sentence - perhaps the "and" at the end of this line needs to be deleted?

File: projects/gloo/api/v1/enterprise/options/extproc/extproc.proto [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+  // timeout when the filter is running in asynchronous mode. Zero is a valid
+  // config which means the timer will be triggered immediately. If not
+  // configured, default is 200 milliseconds.
+  google.protobuf.Duration message_timeout = 5 [(validate.rules).duration = {
+    lte {seconds: 3600}
+    gte {}
+  }];
+
+  // Optional additional prefix to use when emitting statistics. This allows to distinguish
+  // emitted statistics between configured *ext_proc* filters in an HTTP filter chain.
```
Comment (ashishb-solo) [inline projects/gloo/api/v1/enterprise/options/extproc/extproc.proto:63]: Suggestion: "This allows distinguishing statistics emitted by multiple ext proc filters..."

File: projects/gloo/api/v1/enterprise/options/extproc/extproc.proto (hunk 3 above)
Comment (jenshu) [reply projects/gloo/api/v1/enterprise/options/extproc/extproc.proto:49]: haha some of these descriptions are copied directly from the envoy proto. yeah it looks like the `and` should be deleted

**Comment by jenshu** [comment]
undraft so ci can run

File: projects/gloo/pkg/utils/merge.go [hunk 5]
```diff
@@ -0,0 +1,42 @@
+package utils
+
+import (
+	"reflect"
+)
+
+// ShallowMerge sets dst to the value of src, if src is non-zero and dst is zero-valued or overwrite=true.
+func ShallowMerge(dst, src reflect.Value, overwrite bool) {
```
Comment (nfuden) [inline projects/gloo/pkg/utils/merge.go:8]: should we be using mergo given its already polluted our repo? Or is this all we need?

Comment (nfuden) [reply projects/gloo/pkg/utils/merge.go:8]: nevermind I see this is a export of existing functionality!

**Review by nfuden — APPROVED** [review]

**Comment by nfuden** [comment]
Nice stuff! I think ext_proc is a nice escape valve to quickly if inefficiently add new functionality to envoy!

**Pull request merged**